Murano kubernetes: scaleNodesUp/Down and scaleGatewaysUp/Down

tkuterina opened the pull request:
Murano kubernetes tests for check actions scaleNodesUp/Down and scaleGatewaysUp/Down

tkuterina committed:
Murano kubernetes tests for check actions
scaleNodesUp/Down and scaleGatewaysUp/Down

gdyuldin commented on mos_tests/environment/os_actions.py:
```diff
@@ -83,6 +89,11 @@ def __init__(self, controller_ip, user='admin', password='admin',
+
+        murano_endpoint = self.session.get_endpoint(
+            service_type='application_catalog', endpoint_type='publicURL')
+        self.murano = MuranoClient(endpoint=murano_endpoint, token=token,
```
If MuranoClient support initialization with session, let's use it (like GlanceClient, for example)

tkuterina replied:
For muranoclient it is nessesary to use endpoint as for heat

gdyuldin commented on mos_tests/environment/os_actions.py:
```diff
@@ -735,3 +746,170 @@ def rebuild_server(self, server, image):
+    def rand_name(self, name):
+        return name + '_' + str(random.randint(1, 0x7fffffff))
+
+    def create_service(self, environment, session, json_data, to_json=True):
```
`session` isn't used here

tkuterina replied:
session is used in line 756: session_id=session.id

gdyuldin commented on mos_tests/environment/os_actions.py:
```diff
@@ -735,3 +746,170 @@ def rebuild_server(self, server, image):
+        if to_json:
+            service = service.to_dict()
+            service = json.dumps(service)
+            return yaml.load(service)
```
`return service.to_dict()`?

gdyuldin commented on mos_tests/environment/os_actions.py:
```diff
@@ -735,3 +746,170 @@ def rebuild_server(self, server, image):
+    def wait_for_environment_deploy(self, environment):
+        start_time = time.time()
+        status = self.murano.environments.get(environment.id).status
+        while status != 'ready' and time.time() - start_time < 3800:
```
`wait` function

gdyuldin commented:
Let's make MuranoActions for only murano-specific helpers

gdyuldin commented on mos_tests/environment/os_actions.py:
```diff
@@ -735,3 +746,170 @@ def rebuild_server(self, server, image):
+    def get_action_id(self, environment, name, service):
+        env_data = environment.to_dict()
+        a_dict = env_data['services'][service]['?']['_actions']
+        for action_id, action in a_dict.iteritems():
```
iteritems is absent in python3

gdyuldin commented on mos_tests/murano/test_kubernetes.py:
```diff
@@ -0,0 +1,222 @@
+import uuid
+
+
+@pytest.mark.parametrize('cluster', [(1, 2, 1, 2)], indirect=['cluster'])
```
May be better use dict, for better readability?

gdyuldin commented:
Also test scenarios look too similar. May be better to make some refactoring of tests?

tkuterina committed:
Murano kubernetes tests for check actions
ScaleNodesUp and scaleGatewaysUp for Kubernetes
Cluster if maximum nodes limit is already reached

TimurNurlygayanov commented on mos_tests/murano/actions.py:
```diff
@@ -0,0 +1,210 @@
+                return True
+            except RuntimeError:
+                time.sleep(10)
+        assert 0, 'Containers are not ready'
```
I don't like this string. if we want to raise exception here let's use 'raise' then.

TimurNurlygayanov commented on mos_tests/murano/actions.py:
```diff
@@ -0,0 +1,210 @@
+                        assert self.check_port_access(ip, port)
+                        assert self.check_k8s_deployment(ip, port)
+                else:
+                    assert 0, "Instance {} doesn't have floating IP"\
```
please do not use assert 0, it is bad practice.

tkuterina committed:
fix issues after comments

tkuterina committed:
update after comments

tkuterina committed:
update after comments

tkuterina committed:
add package fixture and update of some methods

tkuterina committed:
fix issue

tkuterina committed:
Added test_kub_nodes_down_if_one_present
QA-2442 - Use ScaleNodesDown/ScaleGatewaysDown action if only one minion/gateway is present

gdyuldin commented on mos_tests/murano/actions.py:
```diff
@@ -0,0 +1,238 @@
+                                                    session_id=session.id)
+        if to_json:
+            service = service.to_dict()
+            service = json.dumps(service)
```
You do this: _dict -> string -> dict_
What goal of this transformation is?

gdyuldin commented on mos_tests/murano/actions.py:
```diff
@@ -0,0 +1,238 @@
+class MuranoActions(object):
+    """Murano-specific actions"""
+
+    def __init__(self, os_conn):
```
Let's initialize murano client in this class

tkuterina replied:
done

gdyuldin commented on mos_tests/murano/actions.py:
```diff
@@ -0,0 +1,238 @@
+    def deploy_environment(self, environment, session):
+        self.os_conn.murano.sessions.deploy(environment.id, session.id)
+        try:
+            wait(lambda: self.os_conn.murano.environments.get(environment.id).
```
Even if environment deploy failed, your code will wait ready status for 1000 seconds.
Let's rewrite it like below example

``` python
def is_murano_env_deployed():
    status = self.os_conn.murano.environments.get(environment.id).status
    if status == 'error':
        raise Exception(...)
    return status == 'ready'

wait(is_murano_env_deployed, ...)
```

gdyuldin commented on mos_tests/murano/actions.py:
```diff
@@ -0,0 +1,238 @@
+    def run_action(self, environment, action_id):
+        self.os_conn.murano.actions.call(environment.id, action_id)
+        try:
+            wait(lambda: self.os_conn.murano.environments.get(environment.id).
```
Same, as above

gdyuldin commented on mos_tests/murano/actions.py:
```diff
@@ -0,0 +1,238 @@
+                ports = configuration[2:]
+                ip = self.get_k8s_ip_by_instance_name(environment, inst_name,
+                                                      service_name)
+                if ip and ports and negative:
```
``` python
if ip:
    for port in ports:
        assert self.check_port_access(ip, port, negative)
        assert self.check_k8s_deployment(ip, port, negative)
```

This cover both variants (with negative ==  True and negative == False)

gdyuldin commented on mos_tests/murano/actions.py:
```diff
@@ -0,0 +1,238 @@
+    def check_port_access(self, ip, port, negative=False):
+        result = 1
+        start_time = time.time()
+        while time.time() - start_time < 600:
```
wait

gdyuldin commented on mos_tests/murano/actions.py:
```diff
@@ -0,0 +1,238 @@
+
+    def check_k8s_deployment(self, ip, port, timeout=3600, negative=False):
+        start_time = time.time()
+        while time.time() - start_time < timeout:
```
wait

tkuterina committed:
update after comments

gdyuldin commented on mos_tests/murano/actions.py:
```diff
@@ -0,0 +1,226 @@
+                raise Exception('Environment deploy finished with errors')
+            return status == 'ready'
+
+        try:
```
You don't need to use try ... except here

gdyuldin commented on mos_tests/murano/actions.py:
```diff
@@ -0,0 +1,226 @@
+            sock.close()
+            return result == 0
+        try:
+            wait(connect, timeout_seconds=300,
```
In negative case this code will wait for 5 minutes. Is this behavior is what you want?
